app/controllers/SilhouetteController.scala

```diff
@@ -1,27 +1,101 @@
 package controllers
 
-import com.mohiva.play.silhouette.api.actions.{ SecuredActionBuilder, UnsecuredActionBuilder }
+import com.mohiva.play.silhouette.api._
+import com.mohiva.play.silhouette.api.actions._
 import com.mohiva.play.silhouette.api.repositories.AuthInfoRepository
 import com.mohiva.play.silhouette.api.services.{ AuthenticatorService, AvatarService }
 import com.mohiva.play.silhouette.api.util.{ Clock, PasswordHasherRegistry }
-import com.mohiva.play.silhouette.api.{ EventBus, Silhouette }
 import com.mohiva.play.silhouette.impl.providers.{ CredentialsProvider, GoogleTotpProvider, SocialProviderRegistry }
 import javax.inject.Inject
 import models.services.{ AuthTokenService, UserService }
 import play.api.Logging
 import play.api.http.FileMimeTypes
-import play.api.i18n.{ I18nSupport, Langs, MessagesApi }
+import play.api.i18n.{ Langs, MessagesApi }
 import play.api.libs.mailer.MailerClient
 import play.api.mvc._
-import utils.auth.DefaultEnv
 
+import scala.concurrent.{ ExecutionContext, Future }
 import scala.concurrent.duration.FiniteDuration
 
-abstract class SilhouetteController(override protected val controllerComponents: SilhouetteControllerComponents)
-  extends MessagesAbstractController(controllerComponents) with SilhouetteComponents with I18nSupport with Logging {
+import scala.language.higherKinds
 
-  def SecuredAction: SecuredActionBuilder[EnvType, AnyContent] = controllerComponents.silhouette.SecuredAction
-  def UnsecuredAction: UnsecuredActionBuilder[EnvType, AnyContent] = controllerComponents.silhouette.UnsecuredAction
+abstract class SilhouetteController[E <: Env](
+  override protected val controllerComponents: SilhouetteControllerComponents[E])
+  extends MessagesAbstractController(controllerComponents) with SilhouetteComponents[E] with Logging {
+
+  type SecuredEnvRequest[A] = SecuredRequest[EnvType, A]
```

Comment:
But the subclass has also the type parameter. On the first glance it's confusing where the EnvType comes from. I think if you hard code the type in the SilhouetteComponents and remove the type param from the subclass hierarchy, then this makes sense. But in this example every subclass, has the `Env` type as type param.

Reply:
Good point.  I've taken `EnvType` out as it's not needed.